### ISSUE TITLE
lang: allow `token::..` and `mint::...` to be used as checks without `init`

### ISSUE THREAD

**Issue by paul-schaaf** [project-serum/anchor#1222]

**Comment by blockchainlover2019** [comment]
@paul-schaaf, @armaniferrante, my ticket (remove rent sysvar in token init) has blocked by fanatid.
I agree with fanatid. and Let me do this ticket. I also want this function to be implemented in Anchor.

**Comment by blockchainlover2019** [comment]
Ok. Accepted.

**Comment by blockchainlover2019** [comment]
Hi, @fanatid, Is it good idea to allow token::authority or token::mint? not only both.
```
#[account(
  token::mint = mint
)]
pub token_a : Accout<'info, TokenAccount>

#[account(
  token::authority = payer
)]
pub token_b : Accout<'info, TokenAccount>
```
or only have to mention both
```
#[account(
  token::mint = mint,
  token::authority = payer
)]
pub token_a : Accout<'info, TokenAccount>
```

**Comment by paul-schaaf** [comment]
they have to exist when used with `init`. They shouldnt have to both exist if used without `init`. I think its possible there are situations where you want to save a token account of a particular mint but dont care about its authority

**Comment by blockchainlover2019** [comment]
Yeah, I think so!

**Comment by blockchainlover2019** [comment]
Hi, @paul-schaaf, I implemented all functions for token constraint.
But I have some problem now.
This is test example.
```
#[derive(Accounts)]
pub struct TestConstraintToken<'info> {
    #[account(
        token::mint = mint, 
        token::authority = payer
    )]
    pub token: Account<'info, TokenAccount>,
    pub mint: Account<'info, Mint>,
    #[account(signer)]
    pub payer: AccountInfo<'info>,
    pub rent: Sysvar<'info, Rent>,
    pub system_program: AccountInfo<'info>,
    pub token_program: AccountInfo<'info>,
}
```
```
To deploy this program:
  $ solana program deploy /root/anan/anchor/tests/misc-anan/target/deploy/misc.so
The program address will default to this keypair (override with --program-id):
  /root/anan/anchor/tests/misc-anan/target/deploy/misc-keypair.json
thread 'main' panicked at 'Code not parseable: Error("init must be provided before token")', lang/syn/src/idl/file.rs:348:58
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
I get **Error("init must be provided before token")**, but I already commented this part.
[lang/syn/src/parser/accounts/constraints.rs#L733](https://github.com/project-serum/anchor/blob/9fea72e701813662a4af3733cb14ab713ebfdbef/lang/syn/src/parser/accounts/constraints.rs#L733)
```
  fn add_token_mint(&mut self, c: Context<ConstraintTokenMint>) -> ParseResult<()> {
      if self.token_mint.is_some() {
          return Err(ParseError::new(c.span(), "token mint already provided"));
      }
      if self.associated_token_mint.is_some() {
          return Err(ParseError::new(
              c.span(),
              "associated token mint already provided",
          ));
      }
      // anan/issue#1222
      /*
      if self.init.is_none() {
          return Err(ParseError::new(
              c.span(),
              "init must be provided before token",
          ));
      }*/
      self.token_mint.replace(c);
      Ok(())
  }
```
As you can see, build succeed, and it means it has succeeded in parsing codes.
But the error may come from when parse IDL.
I found "init must be provided before token" this string in all anchor codebase, but there isn't.
What's that? I want to test anchor project but idl parse error occurs.

**Comment by paul-schaaf** [comment]
@blockchainlover2019 pls make a WIP PR and link it to this issue so we can have a look

**Comment by paul-schaaf** [comment]
Are you running `anchor build`? If so, you need to build cli from the `cli` folder and then use that to execute the command. (the anchor cli uses `anchor-lang` for IDL generation but it needs to use the local `anchor-lang`)

I have my local anchor cli bound to an alias
```
alias anchor-local=".../anchor/target/debug/anchor"
```

**Comment by blockchainlover2019** [comment]
Yeah, thanks. I've just figured it out.
I tried to use cli in the project but I wonder how to use it.

**Comment by blockchainlover2019** [comment]
> 
Thanks. Paul. I've just figured it out. And now it works fine.

**Comment by blockchainlover2019** [comment]
@paul-schaaf, I completed. Should I do PR with unit test?
I wrote some additional unit test case in tests/misc.
I want to know your advice If I have something wrong.
Sorry for late. (Recently, busy work made me take no time to do other work)

**Comment by armaniferrante** [comment]
> Should I do PR with unit test? 

Yes that would be great. We can continue discussion in the PR.

**Comment by blockchainlover2019** [comment]
@armaniferrante, I made a PR.
But some tests are failing. I want to know the reason.

**Comment by blockchainlover2019** [comment]
@armaniferrante, Could you move this ticket to **Waiting For Review** region?

**Comment by blockchainlover2019** [comment]
Linked PR: https://github.com/project-serum/anchor/pull/1505

**Comment by paul-schaaf** [comment]
closed by #1505
